persistent-mysql/Database/Persist/MySQL.hs

```diff
@@ -1014,3 +1019,107 @@ mockMigration mig = do
       result = runReaderT . runWriterT . runWriterT $ mig 
   resp <- result sqlbackend
   mapM_ T.putStrLn $ map snd $ snd resp
+
+-- | MySQL specific 'upsert'. This will prevent multiple queries, when one will
+-- do.
+insertOnDuplicateKeyUpdate
+  :: ( PersistEntityBackend record ~ BaseBackend backend
+     , PersistEntity record
+     , MonadIO m
+     , PersistStore backend
+     , backend ~ SqlBackend
+     )
+  => record
+  -> [Update record]
+  -> SqlPersistT m ()
+insertOnDuplicateKeyUpdate record =
+  insertManyOnDuplicateKeyUpdate [record] []
+
+-- | This wraps values of an Entity's 'EntityField', making them have the same
+-- type. This allows them to be put in lists.
+data SomeField record where
+  SomeField :: EntityField record typ -> SomeField record
+
+-- | Do a bulk insert on the given records in the first parameter. In the event
+-- that a key conflicts with a record currently in the database, the second and
+-- third parameters determine what will happen.
+--
+-- The second parameter is a list of fields to copy from the original value.
+-- This allows you to specify that, when a collision occurs, you'll just update
+-- the value in the database with the field values that you inserted.
+--
+-- The third parameter is a list of updates to perform that are independent of
+-- the value that is provided. You can use this to increment a counter value.
+-- These updates only occur if the original record is present in the database.
```

Comment:
It's a little difficult to explain. If you pass something like

```
insertManyOnDuplicateKeyUpdate _ [SomeField UserName] [UserAge +=. 1]
```
then it generates the following clause in the Update:

```
INSERT INTO ... VALUES ...
ON DUPLICATE KEY UPDATE ... 
  `user`.`name` = VALUES(`user`.`name`)
  `user`.`age` = `user`.`age` + 1
```

So the name gets copied from whatever record was being inserted and had a key collision, and the age gets incremented by 1.

Since the `Update` field requires a specific value, that means that we can't make it dependent on the record that we're updating. MySQL would support something like

```sql
ON DUPLICATE KEY UPDATE
  user.age = VALUES(user.age) + 2
```

which would add 2 to the record that hit the duplicate key. There's not a good way to make the update have more complicated values, unfortunately, so this feature would be difficult to incorporate as-is.